docs/ldml/tr35-keyboards.md

```diff
@@ -439,8 +439,202 @@ For example, the sequence `e\m{marker}\u{300}` would be output in NFC as `è`. T
+    const noMarkersNfd = normalizeNFD(noMarkers);
+    if (markerMap.empty()) {
+        // no markers in this segment.
+        return noMarkersNfd;
+    } else if(noMarkersNfd == noMarkers) {
+        // no change in this segment
+        return input;
+    } else {
+        return addBackMarkers(noMarkersNfd, markerMap);
+    }
+}
+
+// represents a removed marker, to be re-added
+interface MarkerEntry {
+    glue: string;      // 'glue' character
+    marker: Marker;  // which marker
+    processed: boolean = false; // track whether this was processed or not
+    eot: boolean = false; // track whether end of text or not
+}
+
+/** parse the string for markers */
+function parseMarkers(s: string): MarkerEntry[] {
+    // Note: parseMarkers() and removeMarkers() (not shown) could be the same function.
+    // queue of markers found
+    let lastMarkers: Marker[] = [];
+    // return value
+    let outputMap: MarkerEntry[] = [];
+
+
+    // iterate string by codepoints
+    for (i = s.begin; i< s.end;) {
+        if (markerAt(i)) {
+            // found a marker - add it to the list
+            lastMarkers.add(markerAt(i));
+            i += /* marker length */;
+        } else {
+            // From 'i' find the glue char for all markers found so far
+            // (in order)
+            // use first char of decomposed string
+            const glue = normalizeNFD(i).codePointAt(0);
+            for (m of markers) {
+                map.add({marker: m, glue: glue});
+            }
+            markers.clear();
+
+            i++; // move to next codepoint
+        }
+    }
+
+    // ran out of codepoints.
+    // any remaining markers are EOT
+    for (m of markers) {
+        map.add({marker: m, eot: true});
+    }
+}
+
+/**
+ * re-add markers to input. Mutates markerMap.
+ * @param input normalized string to re-add markers
+ * @param markerMap list of glued markers. will be mutated
+ */
+function addBackMarkers(input : string, map: MarkerEntry[]) : stringWithMarkers {
+    // quick check
+    if(map.empty()) return input;
+
+    // output string
+    let out = '';
+
+    // first, add all of the EOT entries
+    while (!map.empty() && map[map.length - 1].eot) {
+        const m = map.popEnd(); // remove the last entry
+        out = m.marker + out; // prepend the marker
+    }
+
+    // now, process input codepoints from end to beginning
+    for (i=input.end;i>=input.begin;i--) {
+        // 0. prepend this codepoint
+        out = i + out;
+
+        // 1. any markers at the end of list which match get added
+        while (!map.empty() && map[map.length - 1].glue == i) {
+            // remove last entry, so we don't have to process it any more.
+            const m = map.popEnd();
+            if (m.processed == false) {
+                out = m.marker + out; // prepend marker
+            } // else we already processed it
+        }
+
+        // 2. look for any out of order markers, not right at the end
```

Comment:
I don't understand the purpose of this section, despite the comment.  Could you give an example case where we'd need this extra section?  I thought that part of the point of the `parseMarkers` section was that they'd always be in order.  If so, this section shouldn't be needed.

Reply:
This code looks for the glue characters and inserts the markers.

parseMarkers was called before normalization. 

At this line of code, the glue characters were out of order because of normalization. See the examples, U+0065 U+0302 U+0320 is now U+0065 **U+0320** U+0302.

We first try them in order, because that's the common case. But if the glue characters were rearranged, then look for them out of order.

Reply:
Oh, right.  🤦 That _would_ be a major reason we're doing this.  

Guess I neglected that during the review in the midst of all the other moving parts.

And that'd be why the loop doesn't get broken.  Okay, that was the piece I was missing; the section makes total sense now.  (Even if the pseudocode's been removed for now.